rfcs/20200710-reduction/README.md

```diff
@@ -0,0 +1,220 @@
+| Prod      | reduce(lambda a, b: a * b, x)              | x          | x       | x        |
+| Mean      | sum(x) / n                                 | x          | x       | x        |
+| Variance  | Mean(pow([x_i - Mean(x) for x_i in x], 2)) | x          | x       |          |
+| Std       | pow(Variance(x), 0.5)                      | x          | x       |          |
+| LogSumExp | log(sum([exp(x_i - max(x)) for x_i in x])) + max(x) | x | x       |          |
+| Median    |                                            |            | x       |          |
+| Mode      |                                            |            | x       |          |
+| p-norm    | pow(sum([pow(abs(x_i), p) for x_i in x]), 1/p) | x      | x       |          |
+
+Note: In this RFC Python syntax is used for math formulas
+
+The following algorithms are out of scope due to their low priority:
+* And;
+* Or;
+* Median;
+* Mode.
+
+Also there are normalize operations, which variate between frameworks:
+* OpenVino:
+  ``` python
+  def l2_normalize(x, eps_mode, eps):
+      l2_norm = pow( eps_mode( sum([pow(x_i, 2) for x_i in x]), eps), 0.5)
+      return [x_i / l2_norm for x_i in x]
+  ```
+  where eps_mode = {add, max}
+* TensorFlow:
+  ``` python
+  def l2_normalize(x, eps):
+      l2_norm = pow( max( sum([pow(x_i, 2) for x_i in x]), eps), 0.5)
+      return [x_i / l2_norm for x_i in x]
+  ```
+* PyTorch:
+  ``` python
+  def l2_normalize(x, eps):
+      l2_norm = max( pow( sum([pow(x_i, 2) for x_i in x]), 0.5), eps)
+      return [x_i / l2_norm for x_i in x]
+  ```
+
+# Proposal
+
+## Common part: Dst memory descriptor.
+Dst memory descriptor is required to indicate reduction dimensions and dst data
+type. As a result dst memory descriptor also contains information about dst
+memory layout. Ideally reduction primitive should preserve memory layout as most
+oneDNN primitives to minimize amount of reorders between multiple layouts.
+However mixing of reduction operation and oneDNN blocked layouts leads to the
+following situations:
+* Let's assume user has memory in nChw16c layout and wants to reduce it over all
+ dimensions. Dst memory descriptor will have 4 dimensions similar to src, but
+ will contain ones in all dimensions. The main issue here is dst layout. Since
+ reduction should preserve layout dst will have nChw16c layout which is
+ illogical, because 16c guarantees that size c divides by 16 and oneDNN
+ internals rely on this. To address this issue dst can be padded to have 16
+ elements.
+* Let's assume user has memory in nChw16c layout and wants to reduce it over
+ dimensions h and w. As before dst memory descriptor will have 4 dimensions
+ similar to src, but will contain {n, c, 1, 1} dimensions. In this case layout
+ can be either nChw16c or nchw or nhwc and really depends on what user wants to
+ do with result.
+
+The following behavior is proposed to address these issues:
+* If any is provided for dst layout primitive preserves layout;
```

Comment:
Maybe put a comment that this would be aligned with binary primitive, that is typically used along with the reduction.
For instance:
``` cpp
l2_norm <- reduction(src, {1, c, 1, 1}) // l2_norm.memory_format ~ nChw16c, with n,h,w=1
src <- binary(div, src, l2_norm) // good (?), binary uses the optimized pass
                                 // as the memory formats are the same
```

Reply:
Thanks, added a comment:

> This behavior would be aligned with binary primitive, that is typically used
> along with the reduction.
>
> For example:
> ``` python
> # l2_norm.memory_format ~ nChw16c, with n, h, w = 1
> l2_norm = reduction(src, {1, c, 1, 1})
> # binary uses the optimized pass as the memory formats are the same
> src = binary(div, src, l2_norm)
> ```

Reply:
Thank you!